### ISSUE TITLE
Mean Reward logged to console

### ISSUE THREAD

**Issue by ScriptBono** [Unity-Technologies/ml-agents#2024]
Let's assume I have a fixed `episode length` of 300 steps, `summary_freq `= 1000 and a `maximum reward` of 1 an agent can achieve in a single episode. If the agent achieves in the first three episodes 0.7, 0.8 and 0.9. The output of the console should be 0.8 since this is the average cumulative reward of the episodes that finished within the first 1000 steps? OR does it also include the reward of the first 100 steps of episode 4?

I'm asking this since even though I defined a maximum reward of 1 in an episode I sometimes get a Mean Reward of > 1 and I wonder how this is possible? Isn't the mean reward supposed to describe the mean reward of only finished episodes during that time period? One reason why I encounter this problem may be that I have many agents linked to a single brain?

**Comment by ervteng** [comment]
Hi @ScriptBono, you are correct, if the maximum reward in the episode is 1, there should be no mean rewards greater than one.

How are you guaranteeing that no episode has a reward greater than 1? The rewards reported by Python correspond to cumulative rewards over an episode.

**Comment by ScriptBono** [comment]
I guarantee in the way that the environment is designed such that an agent can get a positive reward of 0.5 only two times. I'm currently using curriculum learning and I have multiple agents linked to the same brain. I didn't notice this phenomenon before but I will do some more tests to ensure that this is due to my design.

**Comment by ScriptBono** [comment]
So I did some more tests. The pink line uses curriculum learning while the blue one doesn't. As you can clearly see the pink one exceeds the threshold of 1 constantly but is always around 1. Where as the blue one only gets close to 1 but never exceeds. I didn't change anything with the reward function. Where can this problem come from?

![image](https://user-images.githubusercontent.com/4007801/57526904-efc70c00-732e-11e9-8787-8636cc72e0fd.png)

**Comment by roboserg** [comment]
@ervteng there is something weird going on with rewards. For example, in my [environment](https://puu.sh/DqqsK/6d73097c9c.png) an agent gets a reward of `-10` while falling off the platform, code:
```
        if (this.transform.localPosition.y < 0 )
        {
            AddReward(-1f);
            Done();
        }
```

However if you log the rewards it gets punished several times with -1, as if `Done();` is NOT called instantly. As if some amount of steps occur before `Done() `is called.

**Comment by ervteng** [comment]
Hi @roboserg, where are you logging rewards? On Python or on Debug.Log() in Unity? Also, where is this function called, in AgentAction(), Update(), or FixedUpdate()?

@ScriptBono, what parameters are you modifying using Curriculum Learning? It's weird that it only happens when that's enabled, as that only modifies the reset parameters, not the reward system. hmm...

**Comment by roboserg** [comment]
@ervteng so I can now confirm the agent gets punished several times even though Done() is called at the next line. Proof - https://puu.sh/Dquh3/1a9785e691.png or video - https://puu.sh/Dqufs/9eada1fae3.mp4

I am logging the reward in the AgentAction like this - https://puu.sh/DquhI/2e16286c1b.png

I mean am I right the agent should only get -1 as reward for falling down and not -3 ? Its as if the Unity is lagging and not registering the position < 0 immediately.

**Comment by ScriptBono** [comment]
Fixed with same solution as described in #2040.

`if ( ! isDone()) {
}`
